firmware/library/L1_Drivers/dac.hpp

```diff
@@ -1,19 +1,76 @@
+// Dac is a driver for the Digital-To-Analog converter for the LPC40xx chip
 #pragma once
-
-#include <cstdint>
-#include <cstdio>
+#include "L0_LowLevel/LPC40xx.h"
+#include "L1_Drivers/pin_configure.hpp"
 
 class DacInterface
 {
  public:
-    virtual bool Initialize();
-    virtual bool WriteDac(uint16_t dac_output);
-    virtual bool SetVoltage(float voltage);
-    // Sine Demonstration Function
-    virtual void SineDac(uint32_t frequency_hz);
-    // Sawtooth Demonstration Function
-    virtual void SawtoothDac(uint32_t frequency_hz);
-    // Triangle Demonstration Function
-    virtual void TriangleDac(uint32_t frequency_hz);
-    virtual void SetBias(bool bias_level);
+    enum BiasLevel: uint8_t
+    {
+        kBiasHigh  = 0,
+        kBiasLow   = 1
+    };
+    virtual void Initialize(void)                   = 0;
+    virtual bool WriteDac(uint16_t dac_output)      = 0;
+    virtual bool SetVoltage(float voltage)          = 0;
+    virtual void SetBias(BiasLevel bias_level)      = 0;
+};
+class Dac : public DacInterface
+{
+ public:
+    static LPC_DAC_TypeDef * dac_ptr;
+    static LPC_SC_TypeDef * sc_ptr;
+    static LPC_IOCON_TypeDef * iocon_ptr;
+    static constexpr float kVref = 3.3;
+    enum Bit : uint8_t
+    {
+        kClockDividerReg = 0,
+        kBiasReg = 16,
+        kDacOutReg = 6,
+        kEnableDacReg = 16
+    };
+    // PinConfigure led3(PinConfigure::CreatePinConfigure<1, 15>());
+    /* Initialize Dac by setting the clock divider and enabling 
+      The Dac Pin withiin IOCON.  Then set the initial Bias level to 0 */
+    void Initialize(void) override
+    {
+        // Set Clock Divider to 1
+        sc_ptr->PCLKSEL = (1 << kClockDividerReg);
+        // Enable DAC Pin P0.26
+        iocon_ptr->P0_26 = (1 << kEnableDacReg);
```

Comment:
Use the pin configure class, do not use the registers directly. Try to stay away modifying registers as much as possible.

Reply:
I am honestly extremely confused on how to use the pin configure class,  Does pin configure somehow point to the register? such as LPC_IOCON or does it do it in some other way?  Does it not modify the registers?  I have tried looking at I2C and GPIO for it but is there any other ones you suggest for me to understand it better? Like  For LPC_SC->PCLKSEL is this even correlated with a pin and port?  Isn't it a specific register that it is pointed at?

Reply:
> Does pin configure somehow point to the register?
Yes, just like how your DAC class points to the DAC registers.
> such as LPC_IOCON or does it do it in some other way? 
Nope, it uses LPC_IOCON. Just does all of the work for you. It is driver to handle all of that work.
> Does it not modify the registers?
It does. But your driver shouldn't modify the LPC_IOCON register. Why waste time doing that, when there is already a drive that does that for you.
>  I have tried looking at I2C and GPIO for it but is there any other ones you suggest for me to understand it better?
Lets think about PinConfigure as if it were a GPIO. If you had a GPIO object in your class, and you gave it a specific port and pin, you should be able to set it as output and pull it high or low, or set it as an input and read if it is high or low. Going back to PinConfigure, does the same thing in nature, in that it configures pins. so you can set them as pull up, pull down, repeater mode. You can set the pin as DAC function, or GPIO function or a PWM, all depending on the pin in question. Its a driver that configures pins. Look at the methods and see what you can tell it to do.
>  Like For LPC_SC->PCLKSEL is this even correlated with a pin and port?
Nope, that has to do with clocks.
>  Isn't it a specific register that it is pointed at?
PinConfigure points to LPC_IOCON, but only during an actual build. During unit tests, it is pointed to a local storage area, where we can read the side effects of the driver's function calls.